File: Ruby/lib/items/aged_item.rb
# frozen_string_literal: true

require 'items/item'

module Items
  # Represents items whose quality increases over time
  class AgedItem < Item
    def tick
      @gilded_rose.days_remaining -= 1

      return if @gilded_rose.quality >= 50

      @gilded_rose.quality += 1
      @gilded_rose.quality += 1 if @gilded_rose.days_remaining <= 0

      ensure_bounded_quality
    end
  end
end

File: JavaScript/src/items/AgedItem.js
import Item from './Item';

// Represents items whose quality increases over time
export default class AgedItem extends Item {
  tick() {
    this.gildedRose.daysRemaining -= 1;

    if (this.gildedRose.quality >= 50) {
      return;
    }

    this.gildedRose.quality += 1;
    if (this.gildedRose.daysRemaining <= 0) {
      this.gildedRose.quality += 1;
    }

    this.ensureBoundedQuality();
  }
}
